### ISSUE TITLE
Using the whole context in map

### ISSUE THREAD

**Issue by tg44** [taskcluster/json-e#385]
Hy!

I have a json from a service like this;
```
{
	"0x07D6": {
		"Device": "0x07D6",
		"Humidity": 45.89,
		"Endpoint": 1,
		"LinkQuality": 126
	}
}
```

I want to "select" the outer key, or the inner Device key. I have no idea what is the outer key.

What I tried so far is;
```
template: {$map: {$eval: 'a'},each(y): {'${y.key}x': {$eval: 'y.val'}}}
context: {a: {a: 1, b: 2, c: 3}}
result: {ax: 1, bx: 2, cx: 3}
```
and going further;
```
template:
{
	$let: {
		$map: {$eval: a},
		each(y): {asd: '${y.key}x'}
	},
	in: {$eval: asd}
}
context: {a: {a: 1, b: 2, c: 3}}
result: cx
```
Based on this I can list the keys under `a`, but I want to list the keys under the context.
Something like:
```
template: '{$eval: 'identity()'}'
context: {a: 1}
result: {a: 1}
```

My other idea was somehow "wildcard" out the key and read the `Device`, with something like `{$eval: '*.Device'}` ofc this is not implementable because of the possibility of colliding keys, but flatten could have work;
```
template: {$flatten: {a: {b: 1}, c: {d: 3}}}
context: {}
result: [{b: 1}, {d: 3}]
```
Also, keys would be handy (not without the identity/context thing);
```
template: {$keys: {a: 1, b: 2}}
context: {}
result: [a, b]
```

My questions;
 - How would you solve the problem with the current toolset?
   - if it is possible can we document something like this as an example?
 - Can we have the above `identity()`, `$flatten` and `$keys` implemented?

**Comment by djmitche** [comment]
This problem is a little hard to solve in most languages, it seems.  Is there a language where there's a natural way to do this?  Perhaps we could copy its approach.

I'm not sure how you're defining `identity()` -- does it return the context as an object?

It looks like the definition of $flatten you're suggesting is what other languages (Python, JS, at least) call `values`?

```python
>>> list({'a': {'b': 1}, 'c': {'d': 3}}.values())
[{'b': 1}, {'d': 3}]
```

As for `$keys`, that sounds pretty reasonable :)

**Comment by tg44** [comment]
In scala we have a function called identity which is kinda `def identity[A](a: A): A = a`, and it is "useful" in `List(true, false).forall(identity)` situations. (Where forall expects an `A => Boolean` function.)

I think giving back the whole context as-is for further transformation is not a bad idea in general. So `identity()` or `context()` or whatever we name it could simply paste the context as-is for further string/object/array/boolean/numeric processing.

This would open up some other usages (I have no idea if right now context should be an object or not);
```
template: '{$eval: 'identity() * 2'}'
context: 5
result: 10
```

**Comment by djmitche** [comment]
Right now, `context` is always an object, and in fact it's sort of a synthetic object in that it's composed of values provided by layers of $let, the context passed to `render`, and the built-in context values.  So `context` would always return something huge and you'd need to filter out `min` and `max` and `fromNow` and any other context values we added later.  I'm not especially convinced that's worthwhile, but it's certainly something we could implement given suitable convincing :)

**Comment by djmitche** [comment]
In a way, `context()` as described is similar to Python's `locals()`, which is seldom-used and probably a source of bugs when it is used.  And it also makes some optimizations hard, as it's raising something the compiler should be able to reason about (for example, to drop unused values) into runtime data (so those values can't be dropped).

Thinking further on this, I expect you're interested in just the user-supplied context (passed to `render).  When we apply JSON-e in Taskcluster, we tend to have one or two context keys that are objects and have many sub-keys.  For example [hooks](https://docs.taskcluster.net/docs/reference/core/hooks/firing-hooks) define a few top-level context keys that will exist, with the "interesting" data in the `payload` context value.  That, in concert with `$keys` and `$values` and maybe some special-purpose `$singleKey` operator, would probably get you the functionality you want.

**Comment by tg44** [comment]
I use context here as the examples uses it like the [string interpolation](https://github.com/taskcluster/json-e#string-interpolation). Without any implementation details. I think we should somehow refer the "data-input" as a whole. 

As I understand there are right now two inputs to the `jsone` function, the template/transformer and the context/data. I want to refer the context/data without the need to extract/know keys from it. This would open up the possibility to the context/data field to be any json type and not just an object. Also, I think this would open up some more-general approaches to the whole tool. If we agree on that keys and values would be a cool feature, what would be the input of these new things? The whole context/data? And what about nested objects? A value from the input contex/data? Then how would you list the keys of the whole input? If its `{$keys: '${a}'} => ["b", "c"]` I can't get back the `"a"` key from an input like `{a: {b: 1, c: 2}}` bcs I can't address the input object as a whole.

I didn't checked the code yet to go into implementation details, bcs I somewhat bypassed the problem with configuring my datasource to wrap my inner object to an outer object where I know the key and now I get;
```json
{"ZbReceived":{"0x07D6":{"Device":"0x07D6","Humidity":44.65,"Endpoint":1,"LinkQuality":123}}}
```
which can be transformed with;
```json
{
   "$let":{
      "$map":{"$eval":"ZbReceived"},
      "each(y)":{
         "key":"${y.key}",
         "value":{
            "$if":"defined('y.val.ZoneStatusChange')",
            "then":false,
            "else":true
         }
      }
   },
   "in":"${value}"
}
```
With `context()` it would be like;
```json
{"0x07D6":{"Device":"0x07D6","Humidity":44.65,"Endpoint":1,"LinkQuality":123}}
```
```json
{
   "$let":{
      "$map":{"$eval":"context()"},
      "each(y)":{
         "key":"${y.key}",
         "value":{
            "$if":"defined('y.val.ZoneStatusChange')",
            "then":false,
            "else":true
         }
      }
   },
   "in":"${value}"
}
```
But right now I can't use the "flat" input.


A totally another problem was that I cant return with `each(y): {value: '${y.val}'}` bcs of "y.val produced an array or object". Yes, and I wanted that :D

**Comment by djmitche** [comment]
> This would open up the possibility to the context/data field to be any json type and not just an object.

I don't think this makes sense.  Context is the current set of in-scope variables.  It is, by definition, a key/value shaped thing.  The expression evaluator uses it to look up values when it sees identifiers.  And operators like $let and $map generate child contexts with different variables.

> I didn't checked the code yet to go into implementation details, bcs I somewhat bypassed the problem with configuring my datasource to wrap my inner object to an outer object where I know the key and now I get; ..

I think this is the right approach -- it defines a variable (in context) named `ZbReceived`.

I see that the combination of `$let` and `$map` in order to unpack a data structure with a single, unknown key in it is awkward.  Ideas for a general way to accomplish that would be welcome.

> A totally another problem was that I cant return with each(y): {value: '${y.val}'} bcs of "y.val produced an array or object". Yes, and I wanted that :D

`'${...}'` represents string interpolation, and will always produce a string.  I think you wanted `{$eval: 'y.val'}`.

**Comment by tg44** [comment]
> '${...}' represents string interpolation, and will always produce a string. I think you wanted {$eval: 'y.val'}.

Yes, I'm totally dumb sometimes 👍 

> I think this is the right approach 

Right now it worked, bcs the datasource was configurable, but it was pure luck on my side... My other solution would be sth like: `jsone(template, {_originalData: input, ...input})` in the caller side, and document this to the users, which is not really nice but probably a not too bad solution...

> Context is the current set of in-scope variables.

I still don't fully understand why it would be so strange to expose the actual (child)context to the caller. (But I now understand why it would be troubling to allow a non-object "context" as input.)


A bit off but;
```
template: {$if: "defined('a.b')", then: 1, else: 0}
context: {a: {b: 1}}
```
It is not really intuitive what is happening. Possibly it would worth a mention that here the `$if: "a['b']"` is the correct syntax [here](https://github.com/taskcluster/json-e#context), bcs we leave the "context". Or even giving a bigger/complex example like;
```
template:  {$if: "defined('a')", then: {$if: "typeof(a) == 'object' && a['b']", then: "a.b", else: "a"}, else: "not found a"}
context: {a: {b: 1}}
```
